File: problem-of-the-day/day05/Reverse Linked List.cpp
// Iterative
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        if(!head)return nullptr;
        ListNode*n = head;
        ListNode*m = head->next;
        ListNode*r;
        while(n && m){
            r = m->next;
            m->next = n;
            n = m;
            m = r;
        }
        head->next = nullptr;
        return n;
    }
};
// Recursive
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        if(!head)return nullptr;
        else if(!(head->next))return new ListNode(head->val);
        ListNode*ans = reverseList(head->next);
        ListNode*n = ans;
        while(n->next)n=n->next;
        n->next = new ListNode(head->val);
        return ans;
    }
};

File: problem-of-the-day/day05/Reverse Linked List.py
# Iterative
class Solution:
    def reverseList(self, head: ListNode) -> ListNode:
        if not head:
            return None
        n = head
        m = head.next
        while n and m:
            r = m.next
            m.next = n
            n = m
            m = r
        head.next = None
        return n


# Recursive
class Solution:
    def reverseList(self, head: ListNode) -> ListNode:
        if not head:
            return None
        elif not head.next:
            return ListNode(head.val)
        ans = self.reverseList(head.next)
        n = ans
        while n.next:
            n = n.next
        n.next = ListNode(head.val)
        return ans
